### PR TITLE
feat: Implement deduplication of stations by hub in TfLService

### LINKED ISSUE #67: Backend: Add deduplicated query parameter to /tfl/stations endpoint
## Summary

Add optional `deduplicated=true` query parameter to `GET /tfl/stations` endpoint to return hub-grouped stations with aggregated lines, eliminating the need for frontend deduplication logic.

## Context

Issue #54 requires the frontend to display hub stations (e.g., "Seven Sisters") as single entries instead of showing multiple child stations. Currently, the frontend would need to:
1. Fetch all stations
2. Group by `hub_naptan_code`
3. Aggregate lines from all child stations
4. Display deduplicated list

This backend enhancement moves that complexity to the server, where it's easier to maintain and test.

## Proposed Solution

Add `deduplicated` query parameter to `/tfl/stations`:

**Without deduplication (current behavior):**
```bash
GET /tfl/stations
```

Returns all stations including hub children:
```json
[
  {
    "id": "uuid-1",
    "tfl_id": "910GSEVNSIS",
    "name": "Seven Sisters (Rail)",
    "hub_naptan_code": "HUBSVS",
    "hub_common_name": "Seven Sisters",
    "lines": ["weaver"],
    ...
  },
  {
    "id": "uuid-2", 
    "tfl_id": "940GZZLUSVS",
    "name": "Seven Sisters",
    "hub_naptan_code": "HUBSVS",
    "hub_common_name": "Seven Sisters",
    "lines": ["victoria"],
    ...
  }
]
```

**With deduplication (new behavior):**
```bash
GET /tfl/stations?deduplicated=true
```

Returns hub-grouped stations:
```json
[
  {
    "id": "uuid-1",  // ID of first child station
    "tfl_id": "HUBSVS",  // Hub code for use in route segments
    "name": "Seven Sisters",  // Hub common name
    "hub_naptan_code": "HUBSVS",
    "hub_common_name": "Seven Sisters",
    "lines": ["victoria", "weaver"],  // Aggregated from all children
    "latitude": 51.58,  // From any child (same location)
    "longitude": -0.07,
    "last_updated": "2025-11-09T..."  // Most recent update
  },
  {
    "id": "uuid-3",
    "tfl_id": "940GZZLUPCO",  // Non-hub stations returned normally
    "name": "Piccadilly Circus",
    "hub_naptan_code": null,
    "hub_common_name": null,
    "lines": ["piccadilly", "bakerloo"],
    ...
  }
]
```

### Backend Changes

**File:** `backend/app/api/endpoints/tfl.py`

Add query parameter:
```python
@router.get("/stations", response_model=list[StationResponse])
async def get_stations(
    line_id: str | None = None,
    deduplicated: bool = False,  # NEW
    db: AsyncSession = Depends(get_db),
) -> list[StationResponse]:
    """Get all stations, optionally filtered by line and/or deduplicated by hub."""
    service = TflService(db)
    if line_id:
        stations = await service.get_stations_for_line(line_id)
    else:
        stations = await service.get_all_stations()
    
    if deduplicated:
        stations = service.deduplicate_stations_by_hub(stations)
    
    return stations
```

**File:** `backend/app/services/tfl_service.py`

Add deduplication method:
```python
def deduplicate_stations_by_hub(self, stations: list[Station]) -> list[Station]:
    """
    Group stations by hub and return one representative per hub.
    
    For hub stations:
    - tfl_id becomes hub_naptan_code
    - name becomes hub_common_name
    - lines aggregated from all children
    - other fields from first child
    
    Non-hub stations returned unchanged.
    """
    hub_groups: dict[str, list[Station]] = {}
    standalone_stations: list[Station] = []
    
    # Group by hub_naptan_code
    for station in stations:
        if station.hub_naptan_code:
            if station.hub_naptan_code not in hub_groups:
                hub_groups[station.hub_naptan_code] = []
            hub_groups[station.hub_naptan_code].append(station)
        else:
            standalone_stations.append(station)
    
    # Create representative stations for each hub
    result = []
    for hub_code, children in hub_groups.items():
        # Use first child as template
        representative = children[0]
        
        # Aggregate unique lines from all children
        all_lines = set()
        for child in children:
            all_lines.update(child.lines)
        
        # Create virtual station representing the hub
        hub_station = Station(
            id=representative.id,
            tfl_id=hub_code,  # Use hub code instead of child station ID
            name=representative.hub_common_name or representative.name,
            latitude=representative.latitude,
            longitude=representative.longitude,
            lines=sorted(all_lines),  # Aggregated lines
            last_updated=max(child.last_updated for child in children),
            hub_naptan_code=hub_code,
            hub_common_name=representative.hub_common_name,
        )
        result.append(hub_station)
    
    # Add standalone stations
    result.extend(standalone_stations)
    
    # Sort by name for consistent ordering
    return sorted(result, key=lambda s: s.name)
```

## Frontend Benefits

**Without this enhancement:**
- Frontend needs 2-3 helper functions for deduplication
- Complex state management in useTflData
- ~100-150 lines of frontend code
- Complex testing

**With this enhancement:**
- Frontend just fetches `GET /tfl/stations?deduplicated=true`
- Zero deduplication logic
- ~10-20 lines of frontend code
- Simple testing

## Testing Requirements

Coverage: 100% on new backend code

### Unit Tests

```python
async def test_deduplicate_stations_by_hub():
    """Test hub stations are deduplicated with aggregated lines."""
    service = TflService(db)
    
    # Create hub children
    child1 = Station(tfl_id="910GSEVNSIS", hub_naptan_code="HUBSVS", 
                     hub_common_name="Seven Sisters", lines=["weaver"])
    child2 = Station(tfl_id="940GZZLUSVS", hub_naptan_code="HUBSVS",
                     hub_common_name="Seven Sisters", lines=["victoria"])
    
    # Create standalone station
    standalone = Station(tfl_id="940GZZLUPCO", hub_naptan_code=None, 
                         name="Piccadilly Circus", lines=["piccadilly"])
    
    stations = [child1, child2, standalone]
    result = service.deduplicate_stations_by_hub(stations)
    
    assert len(result) == 2  # One hub + one standalone
    
    hub = next(s for s in result if s.hub_naptan_code == "HUBSVS")
    assert hub.tfl_id == "HUBSVS"  # Uses hub code
    assert hub.name == "Seven Sisters"  # Uses hub common name
    assert set(hub.lines) == {"victoria", "weaver"}  # Aggregated
    
    standalone_result = next(s for s in result if s.tfl_id == "940GZZLUPCO")
    assert standalone_result.lines == ["piccadilly"]  # Unchanged
```

### Integration Tests

```python
async def test_get_stations_deduplicated(client, test_user_headers):
    """Test GET /tfl/stations?deduplicated=true returns hub-grouped stations."""
    response = await client.get("/tfl/stations?deduplicated=true", 
                                 headers=test_user_headers)
    assert response.status_code == 200
    stations = response.json()
    
    # Verify Seven Sisters appears once with aggregated lines
    seven_sisters = [s for s in stations if s["name"] == "Seven Sisters"]
    assert len(seven_sisters) == 1
    assert seven_sisters[0]["tfl_id"] == "HUBSVS"
    assert "victoria" in seven_sisters[0]["lines"]
    assert "weaver" in seven_sisters[0]["lines"]
```

## Related Issues

- **Depends on:** #50, #51, #52, #53 (backend hub infrastructure - COMPLETE ✅)
- **Blocks:** #54 (frontend hub display - makes implementation much simpler)
- **Related:** #65 (hub code support in route validation)

## Priority

**Medium** - Significantly simplifies #54 frontend implementation, but #54 can be done without it (just with more complexity).

## Estimated Effort

- Implementation: 1-2 hours
- Testing: 1 hour
- **Total: 2-3 hours**

## Success Criteria

- ✅ `GET /tfl/stations?deduplicated=true` returns hub-grouped stations
- ✅ Hub stations use hub code as `tfl_id`
- ✅ Hub stations show aggregated lines from all children
- ✅ Non-hub stations returned unchanged
- ✅ Backward compatible (default behavior unchanged)
- ✅ 100% test coverage on new code
- ✅ All pre-commit hooks pass

## Notes

- This is an **optional enhancement** - #54 can proceed without it
- However, it provides significant architectural benefits:
  - Backend is source of truth for hub grouping
  - Frontend stays simple and declarative
  - Easier to test and maintain
  - Better separation of concerns
- Recommend implementing this **before** #54 for cleaner architecture

### PR DESCRIPTION
**Pull request by mnbf9rca**
- Added `deduplicate_stations_by_hub` method to TfLService to group hub children into representative stations.
- Introduced helper functions `create_hub_representative` and `group_stations_by_hub` for modularity and testability.
- Updated tests for station resolution helpers to cover new deduplication logic.
- Created integration tests for the new deduplication endpoint in the TfL API.
- Added unit tests for the `deduplicate_stations_by_hub` method to ensure correct functionality and edge cases.
- Created integration tests for the GET /tfl/stations?deduplicated endpoint, covering various scenarios including deduplication behavior, line filtering, and sorting.
- Implemented unit tests for the TfLService.deduplicate_stations_by_hub() method, validating the deduplication logic for mixed, standalone, and hub stations, as well as handling edge cases like empty lists and single hub children.

resolves #67

## Summary by Sourcery

Add hub-level station deduplication feature by grouping child stations into single representatives in the service and API, supported by modular helper functions and comprehensive unit and integration tests

New Features:
- Implement deduplication of stations by hub in TfLService via deduplicate_stations_by_hub
- Add deduplicated query parameter to GET /tfl/stations endpoint to return hub-grouped stations

Enhancements:
- Introduce pure helper functions for station deduplication: group_stations_by_hub, aggregate_station_lines, get_latest_update_time, and create_hub_representative
- Refactor station resolution tests to use create_test_station factory for consistent test data

Documentation:
- Update API endpoint docstrings to describe hub deduplication behavior

Tests:
- Add unit tests for deduplication helpers and TfLService.deduplicate_stations_by_hub method
- Add integration tests for GET /tfl/stations?deduplicated endpoint covering deduplication, line filtering, and sorting

Chores:
- Extend conftest with TestRailwayNetwork and create_test_station factory for shared test data

### CODE REVIEW

File: backend/app/api/tfl.py [hunk 1]
```diff
@@ -74,9 +88,31 @@ async def get_stations(
...
+    # show all lines served by the hub, not just the filtered line.
+    if deduplicated:
+        # Fetch all stations to get complete hub information
+        stations = await tfl_service.fetch_stations(line_tfl_id=None)
+        # Deduplicate to create hub representatives with all lines
+        # Pass line_id so hub representatives use UUID from the line-matched station
+        stations = tfl_service.deduplicate_stations_by_hub(stations, line_filter=line_id)
+        # Then filter by line if requested (keeps only stations serving that line)
+        if line_id:
+            stations = [s for s in stations if line_id in s.lines]
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/api/tfl.py:111]: hubs are rarely more than 2-3 stations, and there are about 100 hubs in the network

File: backend/tests/test_station_resolution_helpers.py [hunk 2]
```diff
@@ -278,3 +199,311 @@ def test_station_resolution_error_base_class(self):
...
+        hub_station2 = create_test_station(
+            "HUB1_CHILD_B", "Hub 1 Beta", ["line2"], hub_naptan_code="HUB1", hub_common_name="Hub Alpha"
+        )
+
+        hub_groups, standalone_stations = group_stations_by_hub([hub_station1, hub_station2])
+
+        assert len(hub_groups) == 1
+        assert "HUB1" in hub_groups
+        assert len(hub_groups["HUB1"]) == 2
+        assert standalone_stations == []
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_station_resolution_helpers.py:255]: resolved in 511235b

File: backend/tests/conftest.py [hunk 3]
```diff
@@ -517,3 +520,170 @@ def settings_fixture() -> Settings:
...
+        )
+
+    @staticmethod
+    def create_standalone_delta() -> Station:
+        """Create standalone Delta station serving line2."""
+        return create_test_station(
+            TestRailwayNetwork.STANDALONE_DELTA_ID,
+            TestRailwayNetwork.STANDALONE_DELTA_NAME,
+            [TestRailwayNetwork.LINE_2],
+        )
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/conftest.py:634]: deferred to #70
